google/cloud/storage/internal/openssl_util.h

```diff
@@ -0,0 +1,240 @@
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L  // Older than version 1.1.0
+#define EVP_MD_CTX_new EVP_MD_CTX_create
+#define EVP_MD_CTX_free EVP_MD_CTX_destroy
+#endif
+
+/**
+ * Helper functions for Base64 and related transcoding.
+ */
+struct OpenSslUtils {
+ public:
+  /**
+   * Encode a string using Base64.
+   */
+  static std::string Base64Encode(const std::string& str) {
+    auto bio_chain = MakeBioChainForBase64Transcoding();
+    int retval = 0;
+
+    while (true) {
+      retval = BIO_write(static_cast<BIO*>(bio_chain.get()), str.c_str(),
+                         static_cast<int>(str.length()));
+      if (retval > 0) break;  // Positive value == successful write.
+      if (!BIO_should_retry(static_cast<BIO*>(bio_chain.get()))) {
+        std::ostringstream err_builder;
+        err_builder << "Permanent error in " << __func__ << ": "
+                    << "BIO_write returned non-retryable value of " << retval;
+        google::cloud::internal::RaiseRuntimeError(err_builder.str());
+      }
+    }
+    // Tell the b64 encoder that we're done writing data, thus prompting it to
+    // add trailing '=' characters for padding if needed.
+    while (true) {
+      retval = BIO_flush(static_cast<BIO*>(bio_chain.get()));
+      if (retval > 0) break;  // Positive value == successful flush.
+      if (!BIO_should_retry(static_cast<BIO*>(bio_chain.get()))) {
+        std::ostringstream err_builder;
+        err_builder << "Permanent error in " << __func__ << ": "
+                    << "BIO_flush returned non-retryable value of " << retval;
+        google::cloud::internal::RaiseRuntimeError(err_builder.str());
+      }
+    }
+
+    // This buffer belongs to the BIO chain and is freed upon its destruction.
+    BUF_MEM* buf_mem;
+    BIO_get_mem_ptr(static_cast<BIO*>(bio_chain.get()), &buf_mem);
+    // Return a string copy of the buffer's bytes, as the buffer will be freed
+    // upon this method's exit.
+    return std::string(buf_mem->data, buf_mem->length);
+  }
+
+  /**
+   * Transform a string in-place, removing trailing occurrences of a character.
+   *
+   * Warning: this was written with the intent of operating on a string
+   * containing ASCII-encoded (8-bit) characters (e.g. removing trailing '='
+   * characters from a Base64-encoded string) and may not function correctly
+   * for strings containing Unicode characters.
+   */
+  static std::string& RightTrim(std::string& str, const char& trim_ch) {
+    str.erase(
+        std::find_if(str.rbegin(), str.rend(),
+                     [&trim_ch](char cur_ch) { return trim_ch != cur_ch; })
+            .base(),
+        str.end());
+    return str;
+  }
+
+  /**
+   * Sign a string with the private key from a PEM container.
+   */
+  static std::string SignStringWithPem(
+      const std::string& str, const std::string& pem_contents,
```

Comment:
Elsewhere in the project we have been using `std::string const&`, please change here (and elsewhere) for consistency with the rest of the project.

Reply:
Done.